state/runtime/system/system_unstaking.go

```diff
@@ -0,0 +1,53 @@
+package system
+
+import (
+	"math/big"
+
+	"github.com/0xPolygon/minimal/types"
+)
+
+// unstakingHandler implements the unstaking logic for the System runtime
+type unstakingHandler struct {
+	s *System
+}
+
+// gas returns the fixed gas price of the unstaking operation
+func (uh *unstakingHandler) gas(_ []byte) uint64 {
+	return 40000
+}
+
+// run executes the system contract unstaking method
+// Unstaking is for the FULL amount staked, no partial support is currently present
+func (uh *unstakingHandler) run(state *systemState) ([]byte, error) {
+	// Grab the unstaking address
+	unstakingAddress := types.StringToAddress(GetOperationsMap()["unstaking"])
+
+	// Grab the address calling the staking method
+	staker := state.contract.Caller
+
+	// Grab the staked balance for the account
+	stakedBalance := state.host.GetStakedBalance(staker)
+
+	// Grab the balance on the unstaking address
+	unstakingAccountBalance := state.host.GetBalance(unstakingAddress)
+
+	// Sanity check
+	// Can't unstake if the balance isn't previously present on the unstaking account
+	// Can't unstake if the value is different from 0
+	if unstakingAccountBalance.Cmp(stakedBalance) < 0 || state.contract.Value.Cmp(big.NewInt(0)) != 0 {
+		return nil, nil
+	}
+
+	// Decrease the staked balance on the unstaking address
+	state.host.SubBalance(unstakingAddress, stakedBalance)
```

Comment:
Just curious question, I remember you didn't change the balance of system contract when an account stake balance. 
1. Is it correct behaviour that you reduce the total staked balance from contract?
2. Is unstakingAddress correct address to reduce balance?

Reply:
The reason we modified the balance for the `unstaking address`, as well as the staked balance in the system contract here, is because this event (`unstaking`) is triggered by sending a transaction to the `unstaking address` with a value of **`0`**.

We didn't need to modify it directly like this in the staking handler, because the flow that previously executes in `executor.go` (specifically the `transfer` method) would deduce the staker's balance, and increase the staking address balance.
This just doesn't happen with unstaking, because the value of the transaction is 0.

Reply:
Understood. Sorry, it's my misunderstanding about first question. Yes, transfer is processed by original flow. When a user wants to stake, the user sends to "1001" (and staking event happens on process). So staked amout is saved to "1001" (, which is not unstaking address). It it correct?

Reply:
I understand what you mean now. I subtracted the balance from the wrong address. It should subtract the balance from `staking address` (1001) which holds the funds, not t`unstaking address` (1002).
Fixing now 👍